pkg/webhook/pod/pods.go

```diff
@@ -169,44 +202,82 @@ func (pc *PodAdmissionControl) admitDeletePods(name, namespace string) *admissio
-	if *ownerStatefulSet.Spec.Replicas <= 1 && ownerStatefulSet.Status.Replicas == 1 {
-		klog.Infof("tc[%s/%s]'s statefulset only have one pod[%s/%s],admit to delete it.", namespace, tcName, namespace, name)
-		return util.ARSuccess()
+	payload := &admitPayload{
+		pod:              pod,
+		controller:       tc,
+		ownerStatefulSet: ownerStatefulSet,
+		pdClient:         pc.pdControl.GetPDClient(pdapi.Namespace(namespace), tcName, tc.IsTLSClusterEnabled()),
 	}
+	return pc.admitDeletePdPods(payload)
+}
 
+func (pc *PodAdmissionControl) processAdmitDeleteTiKVPod(pod *core.Pod, ownerStatefulSet *apps.StatefulSet) *admission.AdmissionResponse {
+	name := pod.Name
+	namespace := pod.Namespace
+	l := label.Label(pod.Labels)
 	payload := &admitPayload{
 		pod:              pod,
-		tc:               tc,
 		ownerStatefulSet: ownerStatefulSet,
-		pdClient:         pc.pdControl.GetPDClient(pdapi.Namespace(namespace), tcName, tc.IsTLSClusterEnabled()),
+	}
+	controllerName, exist := pod.Labels[label.InstanceLabelKey]
+	if !exist {
+		klog.Errorf("pod[%s/%s] has no label: %s", namespace, name, label.InstanceLabelKey)
+		return util.ARSuccess()
 	}
 
-	if l.IsPD() {
-		return pc.admitDeletePdPods(payload)
-	} else if l.IsTiKV() {
+	if l.IsTidbClusterPod() {
+		tcName := controllerName
+		tc, err := pc.operatorCli.PingcapV1alpha1().TidbClusters(namespace).Get(tcName, metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				klog.Infof("tc[%s/%s] had been deleted,admit to delete pod[%s/%s]", namespace, tcName, namespace, name)
+				return util.ARSuccess()
+			}
+			klog.Errorf("failed get tc[%s/%s],refuse to delete pod[%s/%s]", namespace, tcName, namespace, name)
+			return util.ARFail(err)
+		}
+		payload.pdClient = pc.pdControl.GetPDClient(pdapi.Namespace(namespace), tcName, tc.IsTLSClusterEnabled())
+		payload.controller = tc
+		payload.controllerDesc = controllerDesc{
+			name:      tcName,
+			namespace: namespace,
+			kind:      v1alpha1.TiDBClusterKind,
+		}
+		return pc.admitDeleteTiKVPods(payload)
+	} else if l.IsGroupPod() {
+		tgName := controllerName
+		tg, err := pc.operatorCli.PingcapV1alpha1().TiKVGroups(namespace).Get(tgName, metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				klog.Infof("tikvgroup[%s/%s] had been deleted,admit to delete pod[%s/%s]", namespace, tgName, namespace, name)
+				return util.ARSuccess()
+			}
+			klog.Errorf("failed get tikvgroup[%s/%s],refuse to delete pod[%s/%s]", namespace, tgName, namespace, name)
+			return util.ARFail(err)
+		}
+		ownerTcName := tg.Spec.ClusterName
+		tc, err := pc.operatorCli.PingcapV1alpha1().TidbClusters(namespace).Get(ownerTcName, metav1.GetOptions{})
+		if err != nil {
+			// Event if the ownerTC is deleted, we won't delete the tikvgroup pod unless its owner controller is deleted
+			klog.Errorf("failed get tc[%s/%s],refuse to delete pod[%s/%s]", namespace, ownerTcName, namespace, name)
+			return util.ARFail(err)
+		}
```

Comment:
If we do not care about tc, why do we check the tc here? Is it OK to remove this block?

Reply:
I think we can't. we need to fetch some cluster level information in tidbcluster spec.